File: fabric-sdk-go/grant/macaroon.go
package grant

import (
	"encoding/json"
	"errors"
	"fmt"

	"gopkg.in/macaroon.v2"
)

// MacaroonVersion is the libmacaroon wire version used by the fabric protocol.
// fabric-spec-001-v1.0.md §"Macaroon structure" pins this to v2.
const MacaroonVersion = macaroon.V2

// ErrSignatureInvalid is returned when macaroon HMAC verification fails.
var ErrSignatureInvalid = errors.New("grant: macaroon signature verification failed")

// MintSpec is the input to Mint.
type MintSpec struct {
	// RootKey is the HMAC secret known to the issuer and verifier. 32 bytes recommended.
	RootKey []byte
	// Location is bound into the macaroon. Use "*" for transport-agnostic Grants.
	Location string
	// Identifier carries the structured grant header that consumers can inspect.
	Identifier Identifier
	// Caveats is the ordered list of first-party caveat strings (per the spec's
	// caveat catalog, e.g. "time < 2026-06-15T18:00:00Z"). For M1 the wrapper
	// passes these through verbatim; semantic checking is the verifier's job.
	Caveats []string
}

// Mint constructs a macaroon from spec and returns the high-level Grant.
// The Grant's Macaroon field holds the libmacaroon v2 wire bytes.
func Mint(spec MintSpec) (*Grant, error) {
	idJSON, err := json.Marshal(spec.Identifier)
	if err != nil {
		return nil, fmt.Errorf("grant.Mint: marshal identifier: %w", err)
	}
	m, err := macaroon.New(spec.RootKey, idJSON, spec.Location, MacaroonVersion)
	if err != nil {
		return nil, fmt.Errorf("grant.Mint: macaroon.New: %w", err)
	}
	for _, c := range spec.Caveats {
		if err := m.AddFirstPartyCaveat([]byte(c)); err != nil {
			return nil, fmt.Errorf("grant.Mint: add caveat %q: %w", c, err)
		}
	}
	bin, err := m.MarshalBinary()
	if err != nil {
		return nil, fmt.Errorf("grant.Mint: marshal: %w", err)
	}
	return &Grant{
		GrantID:    spec.Identifier.GrantID,
		Macaroon:   bin,
		Identifier: spec.Identifier,
		Caveats:    append([]string(nil), spec.Caveats...),
	}, nil
}

// Parse decodes wire-format macaroon bytes into a Grant, extracting the
// structured Identifier and the caveat list. Parse does NOT verify the
// signature; call VerifySignature for that.
func Parse(macBytes []byte) (*Grant, error) {
	var m macaroon.Macaroon
	if err := m.UnmarshalBinary(macBytes); err != nil {
		return nil, fmt.Errorf("grant.Parse: unmarshal: %w", err)
	}
	var id Identifier
	if err := json.Unmarshal(m.Id(), &id); err != nil {
		return nil, fmt.Errorf("grant.Parse: identifier JSON: %w", err)
	}
	caveats := make([]string, 0, len(m.Caveats()))
	for _, c := range m.Caveats() {
		if c.VerificationId != nil {
			// Third-party caveats land here in a later milestone (M2/M3).
			caveats = append(caveats, "third-party:"+string(c.Id))
			continue
		}
		caveats = append(caveats, string(c.Id))
	}
	return &Grant{
		GrantID:    id.GrantID,
		Macaroon:   append([]byte(nil), macBytes...),
		Identifier: id,
		Caveats:    caveats,
	}, nil
}

// CheckFunc evaluates a first-party caveat string against application context.
// Return nil if satisfied; return a non-nil error to reject. M1 verification
// passes a no-op CheckFunc (signature only); M2+ wires in the caveat catalog.
type CheckFunc func(caveat string) error

// AlwaysSatisfied is a CheckFunc that accepts every first-party caveat. Useful
// for the M1 gate where we only test wire-level signature interop.
func AlwaysSatisfied(string) error { return nil }

// VerifySignature checks the HMAC chain over the macaroon given the issuer's
// root key. The check func evaluates first-party caveats; pass AlwaysSatisfied
// to verify only the signature chain.
func VerifySignature(macBytes, rootKey []byte, check CheckFunc) error {
	var m macaroon.Macaroon
	if err := m.UnmarshalBinary(macBytes); err != nil {
		return fmt.Errorf("grant.VerifySignature: unmarshal: %w", err)
	}
	if check == nil {
		check = AlwaysSatisfied
	}
	if err := m.Verify(rootKey, func(c string) error { return check(c) }, nil); err != nil {
		return fmt.Errorf("%w: %v", ErrSignatureInvalid, err)
	}
	return nil
}

File: fabric-sdk-js/src/grant/macaroon.js
// Macaroon mint/parse/verify wrapper. Wire-compatible with the Go SDK via
// gopkg.in/macaroon.v2 / js-macaroon (both implement libmacaroon v2).
// Authoritative caveat catalog: docs/specs/fabric-spec-001-v1.0.md.

import macaroonLib from 'macaroon';
import { bytesToBase64, base64ToBytes } from '../util/base64.js';

const { newMacaroon, importMacaroon } = macaroonLib;

/** libmacaroon wire version used by the fabric protocol. */
export const MACAROON_VERSION = 2;

const enc = new TextEncoder();
const dec = new TextDecoder('utf-8', { fatal: true });

/**
 * SignatureInvalidError is thrown when the macaroon HMAC chain does not verify.
 */
export class SignatureInvalidError extends Error {
  constructor(message, cause) {
    super(message);
    this.name = 'SignatureInvalidError';
    this.code = 'grant_invalid';
    if (cause) this.cause = cause;
  }
}

/**
 * Mint a Grant. The identifier object is JSON-encoded as the macaroon's id bytes.
 *
 * @param {{
 *   rootKey: Uint8Array,
 *   location: string,
 *   identifier: object,
 *   caveats?: string[]
 * }} spec
 * @returns {{ grantId: string, macaroon: Uint8Array, identifier: object, caveats: string[] }}
 */
export function mint(spec) {
  if (!(spec.rootKey instanceof Uint8Array)) {
    throw new TypeError('grant.mint: rootKey must be Uint8Array');
  }
  // The js-macaroon identifier carries the issued_by_keypair as a Uint8Array;
  // serializing through JSON would lose that, so dehydrate to base64 first.
  const idObj = dehydrate(spec.identifier);
  const idBytes = enc.encode(JSON.stringify(idObj));
  const m = newMacaroon({
    identifier: idBytes,
    location: spec.location ?? '',
    rootKey: spec.rootKey,
    version: MACAROON_VERSION,
  });
  const caveats = spec.caveats ?? [];
  for (const c of caveats) {
    m.addFirstPartyCaveat(enc.encode(c));
  }
  return {
    grantId: spec.identifier.grant_id,
    macaroon: m.exportBinary(),
    identifier: spec.identifier,
    caveats: [...caveats],
  };
}

/**
 * Parse a wire-format macaroon. Extracts the Identifier JSON without verifying
 * the signature; pair with verifySignature for the full check.
 *
 * @param {Uint8Array} macBytes
 * @returns {{ grantId: string, macaroon: Uint8Array, identifier: object, caveats: string[] }}
 */
export function parse(macBytes) {
  const m = importMacaroon(macBytes);
  const idText = dec.decode(m._exportAsJSONObjectV2 ? identifierBytesV2(m) : identifierBytesV1(m));
  const identifier = rehydrate(JSON.parse(idText));
  const caveats = [];
  // The js-macaroon Macaroon class doesn't expose caveats() the same way Go does;
  // we round-trip via the JSON export form which is wire-defined per libmacaroon.
  const exported = m.exportJSON();
  // v2 JSON shape has caveats under "c": [{i: "...", l?, vid?}, ...]
  const cs = exported.c ?? exported.caveats ?? [];
  for (const c of cs) {
    const cond = c.i ?? c.cid ?? '';
    caveats.push(cond);
  }
  return {
    grantId: identifier.grant_id,
    macaroon: macBytes,
    identifier,
    caveats,
  };
}

/**
 * Verify the macaroon's HMAC chain under rootKey. The check function is called
 * for each first-party caveat string; throw from check to reject. Pass
 * alwaysSatisfied to verify only the signature.
 *
 * @param {Uint8Array} macBytes
 * @param {Uint8Array} rootKey
 * @param {(caveat: string) => void} check
 */
export function verifySignature(macBytes, rootKey, check = alwaysSatisfied) {
  const m = importMacaroon(macBytes);
  try {
    m.verify(rootKey, check, []);
  } catch (e) {
    throw new SignatureInvalidError('macaroon signature verification failed', e);
  }
}

/** A no-op check function used to verify signature only. */
export function alwaysSatisfied() {}

// --- identifier helpers ---
//
// macaroon objects in JS hold "_identifier_bytes" but the field is private.
// importMacaroon → exportJSON gives us a v2 JSON form where the id is base64
// in 'i64' or text in 'i'. We use exportJSON to extract the id bytes.

function identifierBytesV2(m) {
  const obj = m.exportJSON();
  if (obj.i64 != null) return base64ToBytes(obj.i64);
  if (obj.i != null) return enc.encode(obj.i);
  throw new Error('macaroon identifier missing from exportJSON');
}

function identifierBytesV1(m) {
  const obj = m.exportJSON();
  if (obj.identifier != null) return enc.encode(obj.identifier);
  throw new Error('macaroon identifier missing (v1)');
}

// dehydrate / rehydrate: convert Uint8Array fields in the identifier to base64
// strings before JSON-encoding, and back on parse. Matches what Go's json
// package does for []byte fields.
function dehydrate(obj) {
  return JSON.parse(
    JSON.stringify(obj, (_, v) => (v instanceof Uint8Array ? bytesToBase64(v) : v)),
  );
}
function rehydrate(obj) {
  if (obj && typeof obj === 'object') {
    if ('issued_by_keypair' in obj && typeof obj.issued_by_keypair === 'string') {
      obj.issued_by_keypair = base64ToBytes(obj.issued_by_keypair);
    }
  }
  return obj;
}
